---
# Exploratory Data Analysis
---
This is the process of sifting through the data and trying to make sense of the individual columns and the relationships between them.

We are going to use a dataset from www.fueleconomy.gov that provides information about
makes and models of cars from 1984 through 2018. Using EDA we will explore many of the
columns and relationships found in this data.

## Summary statistics
Summary statistics include the mean, quartiles, and standard deviation. The `.describe`
method will calculate these measures on all of the numeric columns in a DataFrame

In [1]:
import numpy as np
import pandas as pd

In [2]:
fueleco = pd.read_csv("./vehicles.csv.zip")
fueleco.head()

/usr/local/lib/python3.7/dist-packages/IPython/core/interactiveshell.py:2718: DtypeWarning: Columns (70,71,72,73,74,76,79) have mixed types.Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


barrels08  barrelsA08  charge120  ...  phevCity  phevHwy  phevComb
0  15.695714         0.0        0.0  ...         0        0         0
1  29.964545         0.0        0.0  ...         0        0         0
2  12.207778         0.0        0.0  ...         0        0         0
3  29.964545         0.0        0.0  ...         0        0         0
4  17.347895         0.0        0.0  ...         0        0         0

[5 rows x 83 columns]

Call individual summary statistics methods such as `.mean`, `.std`, and `.quantile`

In [3]:
fueleco.mean()

barrels08             17.442712
barrelsA08             0.219276
charge120              0.000000
charge240              0.029630
city08                18.077799
city08U                5.040648
cityA08                0.569883
cityA08U               0.416097
cityCD                 0.000560
cityE                  0.225181
cityUF                 0.000975
co2                   72.538989
co2A                   5.543950
co2TailpipeAGpm       17.826864
co2TailpipeGpm       470.704841
comb08                20.323828
comb08U                5.652724
combA08                0.631160
combA08U               0.453725
combE                  0.230912
combinedCD             0.000459
combinedUF             0.000959
cylinders              5.729105
displ                  3.309829
engId               8582.377382
feScore                0.122580
fuelCost08          2242.470781
fuelCostA08           91.335260
ghgScore               0.120866
ghgScoreA             -0.923889
highway08             24.208588
highway0

In [4]:
fueleco.std()

barrels08              4.580230
barrelsA08             1.143837
charge120              0.000000
charge240              0.487408
city08                 6.970672
city08U               10.462790
cityA08                4.297124
cityA08U               4.093495
cityCD                 0.045392
cityE                  3.180894
cityUF                 0.022785
co2                  163.252019
co2A                  55.956932
co2TailpipeAGpm       93.937414
co2TailpipeGpm       122.351021
comb08                 6.882807
comb08U               11.231748
combA08                4.395797
combA08U               4.131767
combE                  3.244463
combinedCD             0.040759
combinedUF             0.022392
cylinders              1.754487
displ                  1.361925
engId              17606.675590
feScore                2.516348
fuelCost08           601.273869
fuelCostA08          479.485802
ghgScore               2.512612
ghgScoreA              0.651017
highway08              7.128070
highway0

In [6]:
fueleco.quantile(q=np.linspace(0, 1, 5)).T

0.00         0.25  ...          0.75          1.00
barrels08            0.06    14.330870  ...     20.115000     47.087143
barrelsA08           0.00     0.000000  ...      0.000000     18.311667
charge120            0.00     0.000000  ...      0.000000      0.000000
charge240            0.00     0.000000  ...      0.000000     12.000000
city08               6.00    15.000000  ...     20.000000    150.000000
city08U              0.00     0.000000  ...      0.000000    150.000000
cityA08              0.00     0.000000  ...      0.000000    145.000000
cityA08U             0.00     0.000000  ...      0.000000    145.083500
cityCD               0.00     0.000000  ...      0.000000      5.350000
cityE                0.00     0.000000  ...      0.000000    122.000000
cityUF               0.00     0.000000  ...      0.000000      0.896000
co2                 -1.00    -1.000000  ...     -1.000000    847.000000
co2A                -1.00    -1.000000  ...     -1.000000    713.000000
co2TailpipeAGpm      0.00     0.000000  ...      0.000000    713.000000
co2TailpipeGpm       0.00   386.391304  ...    535.789474   1269.571429
comb08               7.00    17.000000  ...     23.000000    136.000000
comb08U              0.00     0.000000  ...      0.000000    136.000000
combA08              0.00     0.000000  ...      0.000000    133.000000
combA08U             0.00     0.000000  ...      0.000000    133.266200
combE                0.00     0.000000  ...      0.000000    121.000000
combinedCD           0.00     0.000000  ...      0.000000      4.800000
combinedUF           0.00     0.000000  ...      0.000000      0.888000
cylinders            2.00     4.000000  ...      6.000000     16.000000
displ                0.00     2.200000  ...      4.300000      8.400000
engId                0.00     0.000000  ...   4401.000000  69102.000000
feScore             -1.00    -1.000000  ...     -1.000000     10.000000
fuelCost08         500.00  1850.000000  ...   2500.000000   6850.000000
fuelCostA08          0.00     0.000000  ...      0.000000   3850.000000
ghgScore            -1.00    -1.000000  ...     -1.000000     10.000000
ghgScoreA           -1.00    -1.000000  ...     -1.000000      8.000000
highway08            9.00    20.000000  ...     27.000000    122.000000
highway08U           0.00     0.000000  ...      0.000000    122.000000
highwayA08           0.00     0.000000  ...      0.000000    121.000000
highwayA08U          0.00     0.000000  ...      0.000000    121.200500
highwayCD            0.00     0.000000  ...      0.000000      4.060000
highwayE             0.00     0.000000  ...      0.000000    120.000000
highwayUF            0.00     0.000000  ...      0.000000      0.877000
hlv                  0.00     0.000000  ...      0.000000     49.000000
hpv                  0.00     0.000000  ...      0.000000    195.000000
id                   1.00  9776.000000  ...  29555.000000  39483.000000
lv2                  0.00     0.000000  ...      0.000000     41.000000
lv4                  0.00     0.000000  ...     13.000000     55.000000
phevBlended          0.00     0.000000  ...      0.000000      1.000000
pv2                  0.00     0.000000  ...      0.000000    194.000000
pv4                  0.00     0.000000  ...     91.000000    192.000000
range                0.00     0.000000  ...      0.000000    335.000000
rangeCity            0.00     0.000000  ...      0.000000    333.100000
rangeCityA           0.00     0.000000  ...      0.000000    103.030000
rangeHwy             0.00     0.000000  ...      0.000000    346.900000
rangeHwyA            0.00     0.000000  ...      0.000000     90.550000
UCity                0.00    18.000000  ...     25.475600    224.800000
UCityA               0.00     0.000000  ...      0.000000    207.262200
UHighway             0.00    27.370400  ...     38.600000    182.700000
UHighwayA            0.00     0.000000  ...      0.000000    173.143600
year              1984.00  1991.000000  ...   2010.000000   2

Call the `.describe` method

In [9]:
fueleco.describe(percentiles=np.linspace(0, 1, 4, endpoint=False)).T

count          mean  ...           75%           max
barrels08        39101.0     17.442712  ...     20.115000     47.087143
barrelsA08       39101.0      0.219276  ...      0.000000     18.311667
charge120        39101.0      0.000000  ...      0.000000      0.000000
charge240        39101.0      0.029630  ...      0.000000     12.000000
city08           39101.0     18.077799  ...     20.000000    150.000000
city08U          39101.0      5.040648  ...      0.000000    150.000000
cityA08          39101.0      0.569883  ...      0.000000    145.000000
cityA08U         39101.0      0.416097  ...      0.000000    145.083500
cityCD           39101.0      0.000560  ...      0.000000      5.350000
cityE            39101.0      0.225181  ...      0.000000    122.000000
cityUF           39101.0      0.000975  ...      0.000000      0.896000
co2              39101.0     72.538989  ...     -1.000000    847.000000
co2A             39101.0      5.543950  ...     -1.000000    713.000000
co2TailpipeAGpm  39101.0     17.826864  ...      0.000000    713.000000
co2TailpipeGpm   39101.0    470.704841  ...    535.789474   1269.571429
comb08           39101.0     20.323828  ...     23.000000    136.000000
comb08U          39101.0      5.652724  ...      0.000000    136.000000
combA08          39101.0      0.631160  ...      0.000000    133.000000
combA08U         39101.0      0.453725  ...      0.000000    133.266200
combE            39101.0      0.230912  ...      0.000000    121.000000
combinedCD       39101.0      0.000459  ...      0.000000      4.800000
combinedUF       39101.0      0.000959  ...      0.000000      0.888000
cylinders        38956.0      5.729105  ...      6.000000     16.000000
displ            38958.0      3.309829  ...      4.300000      8.400000
engId            39101.0   8582.377382  ...   4401.000000  69102.000000
feScore          39101.0      0.122580  ...     -1.000000     10.000000
fuelCost08       39101.0   2242.470781  ...   2500.000000   6850.000000
fuelCostA08      39101.0     91.335260  ...      0.000000   3850.000000
ghgScore         39101.0      0.120866  ...     -1.000000     10.000000
ghgScoreA        39101.0     -0.923889  ...     -1.000000      8.000000
highway08        39101.0     24.208588  ...     27.000000    122.000000
highway08U       39101.0      6.712736  ...      0.000000    122.000000
highwayA08       39101.0      0.736452  ...      0.000000    121.000000
highwayA08U      39101.0      0.523423  ...      0.000000    121.200500
highwayCD        39101.0      0.000343  ...      0.000000      4.060000
highwayE         39101.0      0.238526  ...      0.000000    120.000000
highwayUF        39101.0      0.000938  ...      0.000000      0.877000
hlv              39101.0      2.029539  ...      0.000000     49.000000
hpv              39101.0     10.411243  ...      0.000000    195.000000
id               39101.0  19662.541188  ...  29555.000000  39483.000000
lv2              39101.0      1.834812  ...      0.000000     41.000000
lv4              39101.0      6.155930  ...     13.000000     55.000000
pv2              39101.0     13.649574  ...      0.000000    194.000000
pv4              39101.0     33.883711  ...     91.000000    192.000000
range            39101.0      0.500243  ...      0.000000    335.000000
rangeCity        39101.0      0.458375  ...      0.000000    333.100000
rangeCityA       39101.0      0.050978  ...      0.000000    103.030000
rangeHwy         39101.0      0.450392  ...      0.000000    346.900000
rangeHwyA        39101.0      0.046958  ...      0.000000     90.550000
UCity            39101.0     22.789421  ...     25.475600    224.800000
UCityA           39101.0      0.723139  ...      0.000000    207.262200
UHighway         39101.0     33.884375  ...     38.600000    182.700000
UHighwayA        39101.0      1.009562  ...      0.000000    173.143600
year             39101.0   2000.635406  ...   2010.000000   2018.000000
youSaveSpend     39101.0  -3459.572645  ...  -1500.000000  

To get summary statistics on the object columns, use the `.include` parameter

In [10]:
fueleco.describe(include='object').T

count unique                           top   freq
drive       37912      7             Front-Wheel Drive  13653
eng_dscr    23431    545                         (FFS)   8827
fuelType    39101     14                       Regular  25620
fuelType1   39101      6              Regular Gasoline  26931
make        39101    134                     Chevrolet   3900
model       39101   3816               F150 Pickup 2WD    214
mpgData     39101      2                             N  26400
trany       39090     37               Automatic 4-spd  11045
VClass      39101     34                  Compact Cars   5661
guzzler      2335      3                             G   1356
trans_dscr  15047     52                         CLKUP   7809
tCharger     5816      1                             T   5816
sCharger      738      1                             S    738
atvType      3204      8                           FFV   1383
fuelType2    1490      4                           E85   1383
rangeA       1485    216                           290     74
evMotor       644    125                    288V Ni-MH    117
mfrCode      8289     46                           GMX   1255
c240Dscr       53      5              standard charger     42
c240bDscr      51      4           80 amp dual charger     42
createdOn   39101    195  Tue Jan 01 00:00:00 EST 2013  34199
modifiedOn  39101     68  Tue Jan 01 00:00:00 EST 2013  29438
startStop    7405      2                             N   5176

## Column types
Explore the column types.

In [11]:
# Inspect the .dtypes attribute
fueleco.dtypes

barrels08     float64
barrelsA08    float64
charge120     float64
charge240     float64
city08          int64
               ...   
modifiedOn     object
startStop      object
phevCity        int64
phevHwy         int64
phevComb        int64
Length: 83, dtype: object

Summarize the types of columns

In [12]:
fueleco.dtypes.value_counts()

float64    32
int64      27
object     23
bool        1
dtype: int64

Object types may be strings or categorical data, but they could also be numeric-like values that need to be nudged a little so that they are numeric. 
I typically leave integer columns alone. 
I tend to treat them as continuous values. If the values are float values, this indicates that the column could be:
-  All floating-point values with no missing values
-  Floating-point values with missing values
-  Integer values that were missing some values and hence converted to floats

Save some memory by converting these columns to columns that use less memory.

In [14]:
fueleco.select_dtypes(include='int64').describe().T

count          mean           std  ...      50%      75%      max
city08        39101.0     18.077799      6.970672  ...     17.0     20.0    150.0
cityA08       39101.0      0.569883      4.297124  ...      0.0      0.0    145.0
co2           39101.0     72.538989    163.252019  ...     -1.0     -1.0    847.0
co2A          39101.0      5.543950     55.956932  ...     -1.0     -1.0    713.0
comb08        39101.0     20.323828      6.882807  ...     20.0     23.0    136.0
combA08       39101.0      0.631160      4.395797  ...      0.0      0.0    133.0
engId         39101.0   8582.377382  17606.675590  ...    202.0   4401.0  69102.0
feScore       39101.0      0.122580      2.516348  ...     -1.0     -1.0     10.0
fuelCost08    39101.0   2242.470781    601.273869  ...   2250.0   2500.0   6850.0
fuelCostA08   39101.0     91.335260    479.485802  ...      0.0      0.0   3850.0
ghgScore      39101.0      0.120866      2.512612  ...     -1.0     -1.0     10.0
ghgScoreA     39101.0     -0.923889      0.651017  ...     -1.0     -1.0      8.0
highway08     39101.0     24.208588      7.128070  ...     24.0     27.0    122.0
highwayA08    39101.0      0.736452      4.694207  ...      0.0      0.0    121.0
hlv           39101.0      2.029539      5.959735  ...      0.0      0.0     49.0
hpv           39101.0     10.411243     28.167271  ...      0.0      0.0    195.0
id            39101.0  19662.541188  11413.329199  ...  19552.0  29555.0  39483.0
lv2           39101.0      1.834812      4.407887  ...      0.0      0.0     41.0
lv4           39101.0      6.155930      9.698101  ...      0.0     13.0     55.0
pv2           39101.0     13.649574     31.214466  ...      0.0      0.0    194.0
pv4           39101.0     33.883711     45.991687  ...      0.0     91.0    192.0
range         39101.0      0.500243      9.742080  ...      0.0      0.0    335.0
year          39101.0   2000.635406     10.690422  ...   2001.0   2010.0   2018.0
youSaveSpend  39101.0  -3459.572645   3010.284617  ...  -3500.0  -1500.0   5250.0
phevCity      39101.0      0.094703      2.279478  ...      0.0      0.0     97.0
phevHwy       39101.0      0.094269      2.191115  ...      0.0      0.0     81.0
phevComb      39101.0      0.094141      2.226500  ...      0.0      0.0     88.0

[27 rows x 8 columns]

We can see that the city08 and comb08 columns don't go above 150.

In [15]:
np.iinfo(np.int8)

iinfo(min=-128, max=127, dtype=int8)

In [17]:
np.iinfo(np.int16)

iinfo(min=-32768, max=32767, dtype=int16)

In [18]:
fueleco[['city08', 'comb08']].info(memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 39101 entries, 0 to 39100
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype
---  ------  --------------  -----
 0   city08  39101 non-null  int64
 1   comb08  39101 non-null  int64
dtypes: int64(2)
memory usage: 611.1 KB


In [21]:
(
    fueleco[['city08', 'comb08']]
    .assign(
        city08 = fueleco.city08.astype(np.int16),
        comb08 = fueleco.comb08.astype(np.int16),
    )
    .info(memory_usage='deep')
)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 39101 entries, 0 to 39100
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype
---  ------  --------------  -----
 0   city08  39101 non-null  int16
 1   comb08  39101 non-null  int16
dtypes: int16(2)
memory usage: 152.9 KB


In [22]:
fueleco[['make', 'model']].nunique()

make      134
model    3816
dtype: int64

In [23]:
fueleco[['make', 'model']].info(memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 39101 entries, 0 to 39100
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   make    39101 non-null  object
 1   model   39101 non-null  object
dtypes: object(2)
memory usage: 4.9 MB


In [24]:
fueleco[['make']].info(memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 39101 entries, 0 to 39100
Data columns (total 1 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   make    39101 non-null  object
dtypes: object(1)
memory usage: 2.4 MB


In [26]:
(
    fueleco[['make']]
    .assign(
        make = fueleco.make.astype('category')
    )
    .info(memory_usage='deep')
)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 39101 entries, 0 to 39100
Data columns (total 1 columns):
 #   Column  Non-Null Count  Dtype   
---  ------  --------------  -----   
 0   make    39101 non-null  category
dtypes: category(1)
memory usage: 90.4 KB


In [25]:
fueleco[['model']].info(memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 39101 entries, 0 to 39100
Data columns (total 1 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   model   39101 non-null  object
dtypes: object(1)
memory usage: 2.5 MB


In [27]:
(
    fueleco[['model']]
    .assign(
        model = fueleco.model.astype('category')
    )
    .info(memory_usage='deep')
)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 39101 entries, 0 to 39100
Data columns (total 1 columns):
 #   Column  Non-Null Count  Dtype   
---  ------  --------------  -----   
 0   model   39101 non-null  category
dtypes: category(1)
memory usage: 496.6 KB
